core/src/env/engine/on_chain/ext.rs

```diff
@@ -16,58 +16,133 @@
+            #[inline]
+            fn from(return_code: ReturnCode) -> Self {
+                match return_code {
+                    ReturnCode::Success => Ok(()),
+                    $(
+                        ReturnCode::$name => Err(Error::$name),
+                    )*
+                    ReturnCode::UnknownError => Err(Error::UnknownError),
+                }
+            }
+        }
+    };
+}
+define_error_codes! {
+    /// The called function trapped and has its state changes reverted.
+    /// In this case no output buffer is returned.
+    /// Can only be returned from `seal_call` and `seal_instantiate`.
+    CalleeTrapped = 1,
+    /// The called function ran to completion but decided to revert its state.
+    /// An output buffer is returned when one was supplied.
+    /// Can only be returned from `seal_call` and `seal_instantiate`.
+    CalleeReverted = 2,
+    /// The passed key does not exist in storage.
+    KeyNotFound = 3,
+    /// Transfer failed because it would have brought the sender's total balance
+    /// bwlow the subsistence threshold.
+    BelowSubsistenceThreshold = 4,
+    /// Transfer failed for other not further specified reason. Most probably
+    /// reserved or locked balance of the sender that was preventing the transfer.
+    TransferFailed = 5,
+    /// The newly created contract is below the subsistence threshold after executing
+    /// its constructor so no usable contract instance will be created.
+    NewContractNotFunded = 6,
+    /// No code could be found at the supplied code hash.
+    CodeNotFound = 7,
+    /// The account that was called is either no contract (e.g. user account) or is a tombstone.
+    NotCallable = 8,
+}
+
+type Result = core::result::Result<(), Error>;
 
 mod sys {
+    use super::ReturnCode;
+
+    #[link(wasm_import_module = "seal0")]
     extern "C" {
-        pub fn ext_instantiate(
+        pub fn seal_instantiate(
             init_code_ptr: u32,
             init_code_len: u32,
             gas: u64,
-            value_ptr: u32,
-            value_len: u32,
-            input_data_ptr: u32,
-            input_data_len: u32,
-        ) -> u32;
-
-        pub fn ext_call(
+            endowment_ptr: u32,
+            endowment_len: u32,
+            input_ptr: u32,
+            input_len: u32,
+            address_ptr: u32,
+            address_len_ptr: u32,
+            output_ptr: u32,
+            output_len_ptr: u32,
+        ) -> ReturnCode;
+
+        pub fn seal_call(
             callee_ptr: u32,
             callee_len: u32,
             gas: u64,
-            value_ptr: u32,
-            value_len: u32,
-            input_data_ptr: u32,
-            input_data_len: u32,
-        ) -> u32;
-
-        pub fn ext_transfer(
+            transferred_value_ptr: u32,
+            transferred_value_len: u32,
+            input_ptr: u32,
+            input_len: u32,
+            output_ptr: u32,
+            output_len_ptr: u32,
+        ) -> ReturnCode;
```

Comment:
This is unsound. Seal is allowed to **extend** this enum. Older contracts on newer chains then may produce an invalid discriminant insides the enum. You have to use use `u32` everywhere and convert that to the enum using the `Unknown` variant as fallback with your macro.

Reply:
implemented through `RawReturnCode` which is basically a `u32`.